core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -105,20 +108,50 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
     info("Stopped partition state machine")
   }
 
+
+  def handleStateChangeCallback(ctx: scala.Any): Unit = {
+    val latch: Option[CountDownLatch] = ctx.asInstanceOf[Option[CountDownLatch]]
+    latch.get.countDown()
+  }
+
   /**
-   * This API invokes the OnlinePartition state change on all partitions in either the NewPartition or OfflinePartition
-   * state. This is called on a successful controller election and on broker changes
-   */
-  def triggerOnlinePartitionStateChange() {
+    * This API invokes the OnlinePartition state change on all partitions in either the NewPartition or OfflinePartition
+    * state. This is called on a successful controller election and on broker changes
+    * @param async: if true, all partitions are changed asynchronously. In either sync or async case, this function itself
+    *             only returns when all partition changes are complete
+    */
+  def triggerOnlinePartitionStateChange(async: Boolean) {
```

Comment:
Can't we make all calls to this asynchronous? Why do you want to give the option between sync and async?